File: EndGate/EndGate.Core.Net/GameRunner.cs
using System;
using System.Collections.Concurrent;
using EndGate.Core.Net.Loopers;

namespace EndGate.Core.Net
{
    internal class GameRunner
    {
        private static Lazy<GameRunner> _instance = new Lazy<GameRunner>(() => new GameRunner());
        private ConcurrentDictionary<long, TimedCallback> _callbacks;
        private Looper _updateLoop;

        public GameRunner()
        {
            _callbacks = new ConcurrentDictionary<long, TimedCallback>();
        }

        public static GameRunner Instance
        {
            get
            {
                return _instance.Value;
            }
        }

        public Action<int> Register(Game game)
        {
            var updateCallback = CreateAndCacheCallback(game);

            // Try to start the loop prior to adding our games callback.  This callback may be the first, hence the "Try"
            TryLoopStart();

            // Add our callback to the game loop (which is now running), it will now be called on an interval dictated by updateCallback
            _updateLoop.AddCallback(updateCallback);

            // Updating the "updateRate" is an essential element to the game configuration.
            // If a game is running slowly we need to be able to slow down the update rate.
            return CreateUpdateRateSetter(updateCallback);
        }

        public void Unregister(Game game)
        {
            TimedCallback updateCallback;
            _callbacks.TryRemove(game.ID, out updateCallback);
            _updateLoop.RemoveCallback(updateCallback);

            TryLoopStop();
        }

        private void TryLoopStart()
        {
            if (_callbacks.Count == 1)
            {
                _updateLoop = new Looper();
                _updateLoop.Start();
            }
        }

        private void TryLoopStop()
        {
            if (_callbacks.Count == 0 && _updateLoop != null)
            {
                _updateLoop.Dispose();
                _updateLoop = null;
            }
        }

        private TimedCallback CreateAndCacheCallback(Game game)
        {
            var updateCallback = new TimedCallback
            {
                Callback = game.PrepareUpdate
            };

            // Add the callback to the callback cache
            _callbacks.TryAdd(game.ID, updateCallback);

            return updateCallback;
        }

        private Action<int> CreateUpdateRateSetter(TimedCallback callback)
        {
            return (updateRate) =>
            {
                callback.Fps = updateRate;
            };
        }
    }
}

File: EndGate/EndGate.Core.JS/GameRunner.js
var EndGate;
(function (EndGate) {
    /// <reference path="Interfaces/ITyped.ts" />
    /// <reference path="Loopers/Looper.ts" />
    /// <reference path="Loopers/RepaintLooper.ts" />
    /// <reference path="Loopers/TimedCallback.ts" />
    /// <reference path="Game.ts" />
    (function (_) {
        var GameRunner = (function () {
            function GameRunner() {
                this._type = "GameRunner";
                this._updateCallbacks = {
                };
                this._drawCallbacks = {
                };
                this._updateLoop = null;
                this._drawLoop = null;
                this._callbackCount = 0;
            }
            GameRunner.prototype.Register = function (game) {
                var updateCallback = this.CreateAndCacheUpdateCallback(game);
                var drawCallback = this.CreateAndCacheDrawCallback(game);
                this._callbackCount++;
                // Try to start the loop prior to adding our games callback.  This callback may be the first, hence the "Try"
                this.TryLoopStart();
                // Add our callback to the game loop (which is now running), it will now be called on an interval dictated by updateCallback
                this._updateLoop.AddCallback(updateCallback);
                this._drawLoop.AddCallback(drawCallback);
                // Updating the "updateRate" is an essential element to the game configuration.
                // If a game is running slowly we need to be able to slow down the update rate.
                return this.CreateUpdateRateSetter(updateCallback);
            };
            GameRunner.prototype.Unregister = function (game) {
                var updateCallback, drawCallback;
                if(this._updateCallbacks[game.ID]) {
                    updateCallback = this._updateCallbacks[game.ID];
                    drawCallback = this._drawCallbacks[game.ID];
                    this._updateLoop.RemoveCallback(updateCallback);
                    this._drawLoop.RemoveCallback(drawCallback);
                    delete this._updateCallbacks[game.ID];
                    delete this._drawCallbacks[game.ID];
                    this._callbackCount--;
                    this.TryLoopStop();
                }
            };
            GameRunner.prototype.TryLoopStart = function () {
                if(this._callbackCount === 1) {
                    this._updateLoop = new _.Loopers.Looper();
                    this._updateLoop.Start();
                    this._drawLoop = new _.Loopers.RepaintLooper();
                    this._drawLoop.Start();
                }
            };
            GameRunner.prototype.TryLoopStop = function () {
                if(this._callbackCount === 0 && this._updateLoop != null) {
                    this._updateLoop.Dispose();
                    this._updateLoop = null;
                    this._drawLoop.Dispose();
                    this._drawLoop = null;
                }
            };
            GameRunner.prototype.CreateAndCacheUpdateCallback = function (game) {
                var updateCallback = new _.Loopers.TimedCallback(0, function () {
                    game.PrepareUpdate();
                });
                this._updateCallbacks[game.ID] = updateCallback;
                return updateCallback;
            };
            GameRunner.prototype.CreateAndCacheDrawCallback = function (game) {
                var drawCallback = new _.Loopers.LooperCallback(function () {
                    game.PrepareDraw();
                });
                this._drawCallbacks[game.ID] = drawCallback;
                return drawCallback;
            };
            GameRunner.prototype.CreateUpdateRateSetter = function (callback) {
                return function (updateRate) {
                    callback.Fps = updateRate;
                };
            };
            return GameRunner;
        })();
        _.GameRunner = GameRunner;        
    })(EndGate._ || (EndGate._ = {}));
    var _ = EndGate._;
})(EndGate || (EndGate = {}));
var GameRunnerInstance = new EndGate._.GameRunner();
